python/pyarrow/parquet.py

```diff
@@ -104,6 +111,360 @@ def _get_column_indices(self, column_names):
+        else:
+            self._push_pieces(files, part_keys)
+
+    def _should_silently_exclude(self, path):
+        _, tail = path.rsplit(self.pathsep, 1)
+        return tail.endswith('.crc') or tail in EXCLUDED_PARQUET_PATHS
+
+    def _visit_directories(self, level, directories, part_keys):
+        for path in directories:
+            head, tail = _path_split(path, self.pathsep)
+            name, key = _parse_hive_partition(tail)
+
+            index = self.partitions.get_index(level, name, key)
+            dir_part_keys = part_keys + [(name, index)]
+            self._visit_level(level + 1, path, dir_part_keys)
+
+    def _push_pieces(self, files, part_keys):
+        for path in files:
+            piece = ParquetDatasetPiece(path, partition_keys=part_keys)
+            self.pieces.append(piece)
+
+
+def _parse_hive_partition(value):
+    if '=' not in value:
+        raise ValueError('Directory name did not appear to be a '
+                         'partition: {0}'.format(value))
+    return value.split('=', 1)
+
+
+def _path_split(path, sep):
+    i = path.rfind(sep) + 1
+    head, tail = path[:i], path[i:]
+    head = head.rstrip(sep)
+    return head, tail
+
+
+EXCLUDED_PARQUET_PATHS = {'_SUCCESS'}
+
+
+class ParquetDataset(object):
+    """
+    Encapsulates details of reading a complete Parquet dataset possibly
+    consisting of multiple files and partitions in subdirectories
+
+    Parameters
+    ----------
+    path_or_paths : str or List[str]
+        A directory name, single file name, or list of file names
+    filesystem : Filesystem, default None
+        If nothing passed, paths assumed to be found in the local on-disk
+        filesystem
+    metadata : pyarrow.parquet.FileMetaData
+        Use metadata obtained elsewhere to validate file schemas
+    schema : pyarrow.parquet.Schema
+        Use schema obtained elsewhere to validate file schemas. Alternative to
+        metadata parameter
+    split_row_groups
+    validate_schema
+    """
+    def __init__(self, path_or_paths, filesystem=None, schema=None,
+                 metadata=None, split_row_groups=False, validate_schema=True):
+        if filesystem is None:
+            self.fs = LocalFilesystem.get_instance()
+        else:
+            self.fs = filesystem
+
+        self.pieces, self.partitions = _make_manifest(path_or_paths, self.fs)
+
+        self.metadata = metadata
+        self.schema = schema
+
+        self.split_row_groups = split_row_groups
+
+        if validate_schema:
+            self.validate_schemas()
+
+    def validate_schemas(self):
+        open_file = self._get_open_file_func()
+
+        if self.metadata is None and self.schema is None:
+            self.schema = self.pieces[0].open(open_file).schema
+        elif self.schema is None:
+            self.schema = self.metadata.schema
+
+        # Verify schemas are all equal
+        for piece in self.pieces:
+            file_metadata = piece.open(open_file).metadata
```

Comment:
Closing is happening at gc time, there isn't an explicit close method on the `ParquetFile` object yet. We could add one (in a new patch) though